### PR TITLE
Fix stability test

### LINKED ISSUE #2147: Stability test running for v1.1
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
Fix #2147 
Fix the bug in Stability test. 

Currently stability-test is almost stable excluding local-pv pod would sometimes created error.

```sh
stability-gaosongAPP  8:54 AM
Succeed 3 times, then failed: pod:[kube-system/local-volume-provisioner-45j5l] is unavailable,state is CreateContainerError
operator stability test failed
```
### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: tests/cmd/stability/main.go [hunk 1]
```diff
@@ -236,6 +236,8 @@ func run() {
 		time.Sleep(3 * time.Minute)
 		fta.StartNodeOrDie(physicalNode, node)
 		oa.EmitEvent(nil, fmt.Sprintf("StartNode: %s on %s", node, physicalNode))
+		oa.WaitPodOnNodeReadyOrDie(deployedClusters, node)
+		time.Sleep(3 * time.Minute)
```
Comment (cofyc) [inline tests/cmd/stability/main.go]: why sleeping 3 minutes?

File: tests/failover.go [hunk 2]
```diff
@@ -890,6 +891,47 @@ func (oa *operatorActions) addDataToCluster(info *TidbClusterConfig) (bool, erro
 	return true, nil
 }
 
+func (oa *operatorActions) WaitPodOnNodeReadyOrDie(clusters []*TidbClusterConfig, faultNode string) {
+	if err := wait.Poll(1*time.Minute, 60*time.Minute, func() (bool, error) {
```
Comment (cofyc) [inline tests/failover.go]:
```
for each cluster
  for each pod on the node
    if pod is not read
       return false, nil
return true, nil
```

File: tests/failover.go [hunk 3]
```diff
@@ -890,6 +891,47 @@ func (oa *operatorActions) addDataToCluster(info *TidbClusterConfig) (bool, erro
...
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		slack.NotifyAndPanic(fmt.Errorf("failed to check failover"))
+	}
+}
+
+func (oa *operatorActions) WaitPodOnNodeReady(info *TidbClusterConfig, node string) (bool, error) {
```
Comment (cofyc) [inline tests/failover.go]: this is not necessary

File: tests/cmd/stability/main.go (hunk 1 above)
Comment (cofyc) [reply tests/cmd/stability/main.go]: if the problem is `CheckRecoverOrDie` may fail unexpectedly when pods are not ready. after we have waited for the pods to be ready, IMO, this sleep is not necessary.

File: tests/failover.go (hunk 2 above)
Comment (Yisaer) [reply tests/failover.go]: updated.

File: tests/cmd/stability/main.go (hunk 1 above)
Comment (Yisaer) [reply tests/cmd/stability/main.go]: updated and removed.

File: tests/actions.go [hunk 4]
```diff
@@ -195,8 +195,8 @@ type OperatorActions interface {
 	GetNodeMap(info *TidbClusterConfig, component string) (map[string][]string, error)
 	TruncateSSTFileThenCheckFailover(info *TidbClusterConfig, tikvFailoverPeriod time.Duration) error
 	TruncateSSTFileThenCheckFailoverOrDie(info *TidbClusterConfig, tikvFailoverPeriod time.Duration)
-	DeletePDDataThenCheckFailover(info *TidbClusterConfig, tikvFailoverPeriod time.Duration) error
-	DeletePDDataThenCheckFailoverOrDie(info *TidbClusterConfig, tikvFailoverPeriod time.Duration)
+	ReplacePDThenCheckFailover(info *TidbClusterConfig, tikvFailoverPeriod time.Duration) error
+	ReplacePDThenCheckFailoverOrDie(info *TidbClusterConfig, tikvFailoverPeriod time.Duration)
```
Comment (Yisaer) [inline tests/actions.go:199]: I use changing PD image to simulate the failover, deleting PD data seems not stable here.

File: tests/cmd/stability/main.go [hunk 5]
```diff
@@ -146,9 +147,9 @@ func run() {
 		for _, cluster := range clusters {
 			oa.CheckTidbClusterStatusOrDie(cluster)
 			oa.CheckDisasterToleranceOrDie(cluster)
-			go oa.BeginInsertDataToOrDie(cluster)
+			oa.BeginInsertDataToOrDie(cluster)
 		}
-
+		klog.Infof("clusters deployed and checked")
```
Comment (Yisaer) [inline tests/cmd/stability/main.go:152]: Add some logs to help us know the stability test process if it broken.

File: tests/cmd/stability/main.go [hunk 6]
```diff
@@ -234,16 +242,20 @@ func run() {
 		time.Sleep(3 * time.Minute)
 		fta.StartNodeOrDie(physicalNode, node)
 		oa.EmitEvent(nil, fmt.Sprintf("StartNode: %s on %s", node, physicalNode))
+		oa.WaitPodOnNodeReadyOrDie(deployedClusters, node)
 		oa.CheckRecoverOrDie(deployedClusters)
```
Comment (Yisaer) [inline tests/cmd/stability/main.go:246]: After we restart the node, we should wait the pod running, then recover the failover.

File: tests/cmd/stability/main.go [hunk 7]
```diff
@@ -234,16 +242,20 @@ func run() {
...
 
 		// truncate tikv sst file
 		oa.TruncateSSTFileThenCheckFailoverOrDie(clusters[0], 5*time.Minute)
+		klog.Infof("clusters truncate sst file and checked failover")
 
-		// delete pd data
-		oa.DeletePDDataThenCheckFailoverOrDie(clusters[0], 5*time.Minute)
+		// replace pd image
+		oa.ReplacePDThenCheckFailoverOrDie(clusters[0], 5*time.Minute)
+		klog.Infof("clusters pd data deleted and checked failover")
```
Comment (Yisaer) [inline tests/cmd/stability/main.go:258]: we use replacing pd image to simulate the pd failover.

File: tests/actions.go [hunk 8]
```diff
@@ -1037,6 +1039,10 @@ func (oa *operatorActions) getBlockWriterPod(info *TidbClusterConfig, database s
 			RestartPolicy: corev1.RestartPolicyAlways,
 		},
 	}
+	if info.OperatorTag != "e2e" {
+		pod.Spec.Containers[0].ImagePullPolicy = corev1.PullAlways
+	}
+	return pod
```
Comment (Yisaer) [inline tests/actions.go:1045]: We would always use PullAways Policy in non-e2e case.

File: tests/fault.go [hunk 9]
```diff
@@ -592,27 +592,35 @@ func getFaultNode(kubeCli kubernetes.Interface) (string, error) {
...
 
-	if index == 0 {
-		faultNode = nodes.Items[index+1].Name
-	} else {
-		faultNode = nodes.Items[index-1].Name
+	var filterNodes []string
+	for _, node := range nodes.Items {
+		if node.Name != myNode && node.Name != tillerNodeName {
+			filterNodes = append(filterNodes, node.Name)
+		}
```
Comment (Yisaer) [inline tests/fault.go:615]: avoid delete the node which have tiller or stability-job

File: tests/failover.go [hunk 10]
```diff
@@ -207,13 +226,39 @@ func (oa *operatorActions) TruncateSSTFileThenCheckFailover(info *TidbClusterCon
...
+		if err != nil {
+			return false, nil
+		}
+		return false, nil
+	})
+	err = oa.CheckTidbClusterStatus(info)
+	if err != nil {
+		return err
+	}
+	klog.Info("TruncateSSTFileThenCheckFailover success")
```
Comment (Yisaer) [inline tests/failover.go:260]: we should manually clean failure members after we check tikv failover

File: tests/e2e/util/statefulset/statefulset.go [hunk 11]
```diff
@@ -62,3 +67,14 @@ func IsAllDesiredPodsRunningAndReady(c kubernetes.Interface, sts *appsv1.Statefu
 	klog.Infof("desired pods of sts %s/%s are running and ready (%v)", sts.Namespace, sts.Name, actualPodOrdinals.List())
 	return true
 }
+
+// GetPodList gets the current Pods in ss.
+// e2esset.GetPodList(c, sts) would panic while we want return error if something wrong
+func getPodList(c kubernetes.Interface, ss *appsv1.StatefulSet) (*corev1.PodList, error) {
```
Comment (Yisaer) [inline tests/e2e/util/statefulset/statefulset.go:73]: The original `e2esset.GetPodList(c, sts)` would panic if any error happened, we replace it and the newer function would return an error if error happened.

File: tests/cmd/stability/main.go [hunk 12]
```diff
@@ -440,6 +461,7 @@ func newTidbClusterConfig(ns, clusterName string) *tests.TidbClusterConfig {
 			"discovery.image":                cfg.OperatorImage,
 			"tikv.defaultcfBlockCacheSize":   "8GB",
 			"tikv.writecfBlockCacheSize":     "2GB",
+			"pvReclaimPolicy":                "Delete",
```
Comment (Yisaer) [inline tests/cmd/stability/main.go:464]: we set `Delete` as `pvReclaimPolicy` to avoid disk out of quota.

File: tests/actions.go (hunk 8 above)
Comment (cofyc) [reply tests/actions.go:1045]: In which scenarios? If it's not `e2e`, the image will not be changed,  e.g. v1.1.0-rc.2.

Comment (cofyc) [reply tests/actions.go:1045]: ah, I see. It uses `E2EImage` here.

File: tests/actions.go (hunk 4 above)
Comment (weekface) [reply tests/actions.go:199]: Delete PD data is the case that matters, it is not same with changing PD image.

That may be the problem that we should resolve if it is not stable.

Comment (Yisaer) [reply tests/actions.go:199]: got it, I would use Delete PD data back here soon.

**Comment by cofyc** [comment]
you can squash commits into one (`git rebase -i HEAD~54`), then it will be easily cherry-picked by bot automatically

**Pull request merged**